File: src/main/scala/com/yammer/dropwizard/cli/ManagedCommand.scala
package com.yammer.dropwizard.cli

import org.eclipse.jetty.util.component.AggregateLifeCycle
import com.codahale.fig.Configuration
import com.yammer.dropwizard.{Environment, Service}
import com.yammer.dropwizard.jetty.JettyManaged

trait ManagedCommand extends ConfiguredCommand {
  final def run(service: Service, config: Configuration, opts: Map[String, List[String]], args: List[String]): Option[String] = {
    val aggregate = new AggregateLifeCycle
    val env = new Environment
    service.configure(config, env)
    env.jettyObjects.foreach(aggregate.addBean)
    env.managedObjects.map {new JettyManaged(_)}.foreach(aggregate.addBean)
    aggregate.start()
    try {
      run(opts, args)
    } finally {
      aggregate.stop()
    }
  }

  def run(opts: Map[String, List[String]], args: List[String]): Option[String]
}

File: dropwizard-core/src/main/java/com/yammer/dropwizard/cli/ManagedCommand.java
package com.yammer.dropwizard.cli;

import com.yammer.dropwizard.AbstractService;
import com.yammer.dropwizard.config.Configuration;
import com.yammer.dropwizard.config.Environment;
import com.yammer.dropwizard.config.LoggingFactory;
import com.yammer.dropwizard.logging.Log;
import org.apache.commons.cli.CommandLine;

public abstract class ManagedCommand<T extends Configuration> extends ConfiguredCommand<T> {
    private static final Log LOG = Log.forClass(ManagedCommand.class);

    protected ManagedCommand(String name,
                             String description) {
        super(name, description);
    }

    @Override
    protected final void run(AbstractService<T> service,
                             T configuration,
                             CommandLine params) throws Exception {
        new LoggingFactory(configuration.getLoggingConfiguration()).configure();
        final Environment environment = new Environment(configuration, service);
        service.initializeWithBundles(configuration, environment);
        LOG.info("Starting {}", service.getName());
        environment.start();
        try {
            run(configuration, environment, params);
        } finally {
            environment.stop();
        }
    }

    protected abstract void run(T configuration,
                                Environment environment,
                                CommandLine params) throws Exception;
}
